Jib fails the build with long filenames (>100 bytes)

**Description of the issue**:
When building with some long protobuf filenames, Jib breaks with an error that the filename is too long.

**Expected behavior**:
It should build anyways because we all know how long Java filenames can get, especially with paths included.

**Steps to reproduce**:
1) have an app with deep proto import paths
2) use a long name for a proto
3) make sure the combined values are over 100 bytes
4) ask Jib to build an image with them in it

**Environment**:
_Jib:_ `0.1.1-SNAPSHOT`

_Maven:_
```
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T08:41:47-08:00)
Maven home: /usr/local/Cellar/maven/3.3.9/libexec
Java version: 1.8.0_121, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_121.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.13.3", arch: "x86_64", family: "mac"
```

_IntelliJ_:
```
IntelliJ IDEA 2018.1 EAP (Ultimate Edition)
Build #IU-181.3494.3, built on February 6, 2018
IntelliJ IDEA EAP User
Expiration date: March 8, 2018
JRE: 1.8.0_152-release-1136-b5 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.13.3
```

**`jib-maven-plugin` Configuration:**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.sample</groupId>
    <artifactId>java-sample</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <app.version>latest</app.version>
        <app.project>(redacted)</app.project>
        <app.repository>(redacted)</app.repository>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <dependencies>
        <dependency>
            <groupId>com.google.inject</groupId>
            <artifactId>guice</artifactId>
            <version>4.1.0</version>
        </dependency>
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>24.0-jre</version>
        </dependency>
        <dependency>
            <groupId>com.google.protobuf</groupId>
            <artifactId>protobuf-java</artifactId>
            <version>3.5.1</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-netty</artifactId>
            <version>1.9.0</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-protobuf</artifactId>
            <version>1.9.0</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-stub</artifactId>
            <version>1.9.0</version>
        </dependency>
    </dependencies>

    <build>
        <extensions>
            <extension>
                <groupId>kr.motd.maven</groupId>
                <artifactId>os-maven-plugin</artifactId>
                <version>1.5.0.Final</version>
            </extension>
        </extensions>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.7.0</version>
                <configuration>
                    <source>1.8</source>
                    <target>1.8</target>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.xolstice.maven.plugins</groupId>
                <artifactId>protobuf-maven-plugin</artifactId>
                <version>0.5.0</version>
                <configuration>
                    <protocArtifact>com.google.protobuf:protoc:3.5.1-1:exe:${os.detected.classifier}</protocArtifact>
                    <pluginId>grpc-java</pluginId>
                    <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.9.0:exe:${os.detected.classifier}</pluginArtifact>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>compile</goal>
                            <goal>compile-custom</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>com.google.cloud.tools</groupId>
                <artifactId>jib-maven-plugin</artifactId>
                <version>0.1.1-SNAPSHOT</version>
                <configuration>
                    <registry>us.gcr.io</registry>
                    <repository>${app.project}/${app.repository}</repository>
                    <from>us.gcr.io/${app.project}/v2-java:${app.version}</from>
                    <jvmFlags>
                        <jvmFlag>-Xms512m</jvmFlag>
                    </jvmFlags>
                    <mainClass>(redacted)</mainClass>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```

**Log output**:
```
...
[INFO] RUNNING	Pushing new image
[INFO] Pushing new image : 0.172 ms
[INFO] Building and pushing image : 9203.004 ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 32.573 s
[INFO] Finished at: 2018-02-12T14:17:09-08:00
[INFO] Final Memory: 34M/786M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.1.1-SNAPSHOT:build (default-cli) on project platform: Build image failed: file name 'app/classes/com/<really_long_path>/LocationIntegrationSettings$GenericDefaultEntryHolder.class' is too long ( > 100 bytes) -> [Help 1]
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -25,3 +25,4 @@ All notable changes to this project will be documented in this file.
 - Cannot use private base image ([#68](https://github.com/google/jib/pull/68))
 - Building applications with no resources ([#73](https://github.com/google/jib/pull/73))
 - Pushing to registries like Docker Hub and ACR ([#75](https://github.com/google/jib/issues/75))
+- Cannot build with files having long file names (> 100 chars) ([#91](https://github.com/google/jib/issues/91)) 

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarStreamBuilder.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarStreamBuilder.java
@@ -41,6 +41,9 @@ public class TarStreamBuilder {
       List<TarArchiveEntry> entries, OutputStream tarByteStream) throws IOException {
     try (TarArchiveOutputStream tarArchiveOutputStream =
         new TarArchiveOutputStream(tarByteStream)) {
+      // Enables PAX extended headers to support long file names.
+      tarArchiveOutputStream.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
+
       for (TarArchiveEntry entry : entries) {
         tarArchiveOutputStream.putArchiveEntry(entry);
         if (entry.isFile()) {

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/tar/TarStreamBuilderTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/tar/TarStreamBuilderTest.java
@@ -60,6 +60,10 @@ public class TarStreamBuilderTest {
         new TarArchiveEntry(fileA.toFile(), "some/path/to/resourceFileA"));
     testTarStreamBuilder.addEntry(new TarArchiveEntry(fileB.toFile(), "crepecake"));
     testTarStreamBuilder.addEntry(new TarArchiveEntry(directoryA.toFile(), "some/path/to"));
+    testTarStreamBuilder.addEntry(
+        new TarArchiveEntry(
+            fileA.toFile(),
+            "some/really/long/path/that/exceeds/100/characters/abcdefghijklmnopqrstuvwxyz0123456789012345678901234567890"));
   }
 
   @Test
@@ -118,6 +122,15 @@ public class TarStreamBuilderTest {
     // Verifies directoryA was archived correctly.
     TarArchiveEntry headerDirectoryA = tarArchiveInputStream.getNextTarEntry();
     Assert.assertEquals("some/path/to/", headerDirectoryA.getName());
+
+    // Verifies the long file was archived correctly.
+    TarArchiveEntry headerFileALong = tarArchiveInputStream.getNextTarEntry();
+    Assert.assertEquals(
+        "some/really/long/path/that/exceeds/100/characters/abcdefghijklmnopqrstuvwxyz0123456789012345678901234567890",
+        headerFileALong.getName());
+    String fileALongString =
+        CharStreams.toString(new InputStreamReader(tarArchiveInputStream, StandardCharsets.UTF_8));
+    Assert.assertEquals(expectedFileAString, fileALongString);
 
     Assert.assertNull(tarArchiveInputStream.getNextTarEntry());
   }

